### ISSUE TITLE
Blackhole: Unpacker hanging in `unpack_tilizeA_B_block`

### ISSUE THREAD

**Issue by mywoodstock** [tenstorrent/tt-metal#15756]
`unpack_tilizeA_B_block` calls the LLK: `llk_unpack_tilizeA_B_block<true, true>(...)`.

Branch: `asarje/bh-rn50-20241205-test`
Build: `./build_metal.sh --debug`
Run:
``` 
TT_METAL_DPRINT_RISCVS=TR TT_METAL_DPRINT_CORES=0,0 pytest "tests/ttnn/unit_tests/operations/test_maxpool2d.py::test_run_max_pool[dtype=DataType.BFLOAT8_B-dilation=(1, 1)-stride=(2, 2)-padding=(1, 1)-kernel_size=(3, 3)-act_shape=[1, 64, 112, 112]-device_params={'l1_small_size': 24576}]"
```
This test will hang in a call to the above function (works for initial few iterations). I was able to narrow that down using debug printing. 
I have limited the grid to 64 cores to match WH, and I couldn't get smaller grid size as either data doesn't fit, or the issue does not appear.

Relevant kernels:
compute = `ttnn/cpp/ttnn/operations/pool/generic/device/kernels/compute/max_pool_multi_core.cpp`
reader = `ttnn/cpp/ttnn/operations/pool/generic/device/kernels/dataflow/reader_max_pool_2d_multi_core_sharded_with_halo_v2.cpp`

Likely some recent update broke it because this test was working on BH a about week ago.

**Comment by mywoodstock** [comment]
Did some more debug printing, and apart from unpacker handing at the above LLK, the packer seems to be hanging in the previous iteration at `pack_untilize_dst`. Perhaps these are related?

**Comment by ttmtrajkovic** [comment]
@ncvetkovicTT, @nvelickovicTT,
are you aware of any changes to tilize / untilize functions? 

@mywoodstock, is it possible to do git bisect to determine which change caused this to start failing - if you're confident this was indeed passing?

**Comment by nvelickovicTT** [comment]
@ttmtrajkovic there were recent changes in tilize/untilize functions to fix attention matmul kernel, however, it shouldn't cause different behavior between iterations of a kernel. That part is weird, and test being multicore makes it more difficult to debug.

@mywoodstock if you can pinpoint the commit in tt-metal where this started failing, it would help us figure out what went wrong.

**Comment by mywoodstock** [comment]
Found the faulting commit to be `62e8824270`. There is an LLK change to `unpack_tilize` in this commit:
```
diff --git a/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_tilize_api.h b/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_tilize_api.h
index a0aed4572a..42b3965f24 100644
--- a/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_tilize_api.h
+++ b/tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_unpack_tilize_api.h
@@ -52,8 +52,14 @@ inline void llk_unpack_tilize_init(const std::uint32_t operand, const std::uint3
 }

 inline void llk_unpack_tilize_uninit(const std::uint32_t operand, const std::uint32_t face_r_dim = FACE_R_DIM) {
+    // Revert X dim value to default.
     TT_SETADCXX(p_setadc::UNP_A, face_r_dim * FACE_C_DIM - 1, 0x0);
     TT_SETADCXX(p_setadc::UNP_B, face_r_dim * FACE_C_DIM - 1, 0x0);
+
+    // Revert Z dim value back to default.
+    const uint Tile_z_dim = get_operand_num_faces(operand);
+    cfg_reg_rmw_tensix<THCON_SEC0_REG0_TileDescriptor_ADDR32+1, 16, 0xffff0000>(Tile_z_dim);
+
     std::uint32_t operand_id = get_operand_id(operand);
     unpack_config_u config = {0};
```
cc: @bbradelTT

**Comment by bbradelTT** [comment]
That commit was necessary to have attn matmul working properly. I'm pretty sure it fixed a real bug.

**Comment by bbradelTT** [comment]
That commit also updated [llk_lib/llk_unpack_tilize.h](https://github.com/tenstorrent/tt-llk-bh/compare/50796e4f179ed441db96becb59eec17f70a20a7e...b932c3ce7ff67ebe8d94bfbeda4c4ebf7137dee0#diff-e6483556f52a631e97da237fa9d351cd1c731a23814743a21849aea7c8243915)

Regardless, this is strange behaviour.

**Comment by mywoodstock** [comment]
Ah, right there's an update to the `llk_lib` as well, probably some update from that has issues.

**Comment by nvelickovicTT** [comment]
@mywoodstock seems that your branch pulled in BH LLK submodule update, but not corresponding tt-metal update. When I rebased your branch to latest tt-metal main, and rebuilt, the test passes for me. Can you just double check?

**Comment by nvelickovicTT** [comment]
OK, I had an issue with rebase on my side. So, what I actually tested was main only. But that passes. Does your branch modify the test somehow?

**Comment by mywoodstock** [comment]
Yes, I created the test mentioned above in my branch where I reduced the grid size (larger output per core). This test works on full grid (smaller output per core). For using full grid, a larger version will hang. On `main` could you try this one:

```
pytest "tests/ttnn/unit_tests/operations/test_maxpool2d.py::test_run_max_pool[dtype=DataType.BFLOAT8_B-dilation=(1, 1)-stride=(2, 2)-padding=(1, 1)-kernel_size=(3, 3)-act_shape=[8, 64, 112, 112]-device_params={'l1_small_size': 24576}]"
```

**Comment by cmaryanTT** [comment]
@nvelickovicTT - are you unblocked based on Abhinav's feedback?

**Comment by nvelickovicTT** [comment]
@cmaryanTT yes to some extent. It's still tricky to debug a test running on so many cores. Means we can't use the simulator.
Last such hang ended up being debugged for months by more than 10 people in total.
@mywoodstock is 8x8 the smallest grid where you were able to reproduce the issue? Is there a way to modify the test so some data would be reused (if they cannot fit) and still keep the same processing to reproduce the hang?

**Comment by nvelickovicTT** [comment]
@mywoodstock I've managed to reproduce the hang on a 2x1 grid in silicon. I would like to try and simulate that setup, however, simulator only works for SLOW DISPATCH MODE. But, the test fails in SLOW DISPATCH MODE.
Is it possible to fix the test so it runs when in SLOW DISPATCH MODE?

**Comment by nvelickovicTT** [comment]
Created an issue for slow dispatch mode not working - https://github.com/tenstorrent/tt-metal/issues/16191.
Blocked on debugging without it.

**Comment by nvelickovicTT** [comment]
@mywoodstock I have created a PR with the fix for this issue in https://github.com/tenstorrent/tt-llk-bh/pull/62.
I've used your example on main to test. Please give it a try.

Also, In case you need it, I've managed to reduce this test to 1x1 core by setting the activation shape to [1, 16, 2, 16] and [1, 16, 1, 64]. Without the fix above, first configuration passed, and the second hanged.

**Comment by mywoodstock** [comment]
@nvelickovicTT Thanks for the update. I tried running the unit tests with your commits cherry-picked into the bh submodule, and does not hang anymore, so thats great! I do see a slight PCC mismatch with BFLOAT16 case -- but that may or may not be related. I will debug that separately. Please let me know once this PR lands in the BH module. Thanks!

**Comment by nvelickovicTT** [comment]
@mywoodstock PR is merged. Feel free to pull in latest BH LLK main.
